### PR TITLE
prov/sm2: proto-based progress

### PR DESCRIPTION
**Pull request by lrbison**

### CODE REVIEW

File: prov/sm2/src/sm2_progress.c [hunk 1]
```diff
@@ -166,47 +166,63 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
...
-		case ofi_op_tagged:
-			ret = sm2_progress_recv_msg(ep, xfer_entry);
+		switch (xfer_entry->hdr.proto) {
+		case (sm2_proto_return):
+			ret = sm2_progress_proto_return(ep, xfer_entry);
+			break;
+		case (sm2_proto_inject):
+			ret = sm2_progress_proto_inject(ep, xfer_entry);
 			break;
-		default:
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:206]: Why are you getting rid of the default case?

File: prov/sm2/src/sm2_progress.c [hunk 2]
```diff
@@ -166,47 +166,63 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
...
-			continue;
-		}
-
-		switch (xfer_entry->hdr.op) {
-		case ofi_op_msg:
-		case ofi_op_tagged:
-			ret = sm2_progress_recv_msg(ep, xfer_entry);
+		switch (xfer_entry->hdr.proto) {
+		case (sm2_proto_return):
+			ret = sm2_progress_proto_return(ep, xfer_entry);
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:221]: Can we rename this to be inject return?

File: prov/sm2/src/sm2_progress.c [hunk 3]
```diff
@@ -166,47 +166,63 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
...
-				xfer_entry);
-			continue;
-		}
-
-		switch (xfer_entry->hdr.op) {
-		case ofi_op_msg:
-		case ofi_op_tagged:
-			ret = sm2_progress_recv_msg(ep, xfer_entry);
+		switch (xfer_entry->hdr.proto) {
+		case (sm2_proto_return):
```
Comment (a-szegel) [inline prov/sm2/src/sm2_progress.c:220]: Case 1: Progress
```
sm2_progress_recv(): Switch 1 PROTO 
sm2_progress_proto_inject(): Switch 2 OP
sm2_start_common(): Switch 3 PROTO     #### BAD, we shouldn't switch twice on the same thing... this needs to go
```

Case 2 posting buffers:
```
sm2_start_common(): Switch 1 PROTO  # this switch needs to happen on unexpected msg path
```

I think part of this change needs to get rid of sm2_start_common() b/c now those paths are diverging.

**Review by a-szegel — CHANGES_REQUESTED** [review]

**Comment by lrbison** [comment]
Closing this for now.  It might be back with major changes.

**Pull request closed without merging**